Make payment section always visible

ThomazFB opened the pull request:
Summary
==========
Fixes issues #5908 and #5907 by making the payment section always visible and adjusting the Fee percentage mode to be available only when the Order contains any price value as total beside the fee itself. Since both issues scenarios are tied together, I decided to tackle everything in a single PR testing all at once.

Screenshots
==========
https://user-images.githubusercontent.com/5920403/156269729-6f39057f-f3c5-462c-83df-b1242ac83125.mp4

How to Test
==========
1. Make sure the feature flag `ORDER_CREATION_M2` is enabled.
2. Open order creation form.
3. Notice the Payment section is now visible from the start
4. Notice the button "Add fee" on the payment section and click on it.
5. Once inside the fee view, notice the `percentage fee` toggle is gone since there's no valid order total
6. Enter a fee amount and hit `done`.
7. Now back at the New Order view, add a product to the order and enter the `Fee` view back again
8. Notice that the `percentage fee` toggle is now visible


Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Introduce visibility control to Percentage switch inside OrderCreationFeeViewModel

ThomazFB committed:
Add test coverage on Percentage switch visibility control

ThomazFB committed:
Connect percentage switch visibility control with Fee fragment UI

ThomazFB committed:
Make payments section always visible

ThomazFB committed:
Merge branch 'trunk' into issue/5908-hide-fee-percentage

ThomazFB committed:
Remove items rule out of order draft updates

ThomazFB committed:
Make percentage fee mode visibility dependant from order subtotal

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/CreateOrUpdateOrderDraft.kt:
```diff
@@ -20,7 +20,6 @@ class CreateOrUpdateOrderDraft @Inject constructor(
     @OptIn(ExperimentalCoroutinesApi::class)
     operator fun invoke(changes: Flow<Order>, retryTrigger: Flow<Unit>): Flow<OrderDraftUpdateStatus> {
         return changes
-            .filter { it.items.isNotEmpty() }
```
I wonder if we should remove the check completely, or update it to add checks on fees and shipping? I mean something like this:
```kotlin
it.items.isNotEmpty() || it.feeLines.isNotEmpty() || it.shippingLines.isNotEmpty()
```
This would avoid the eager creation, and the odd loading spinner that gets displayed when the form is opened.

ThomazFB replied:
Yes, I also believe this would be the best solution, I didn't like the eager draft posting inside the form, and I wasn't willing to modify the tests for the use case just to consider this scenario. I think we just need a way to make clear in the code that this filter must be incremented as we add more price modifiers inside the form in the future, just like we do with the "isEquivalent".

ThomazFB replied:
Just as a follow-up, changes are done here: https://github.com/woocommerce/woocommerce-android/pull/5956/commits/802002c4ab5b56ea266d7a7e5589d79655c80cfd. I've also moved the PR out of draft state.

ThomazFB committed:
Improve Order changes Flow filter inside CreateOrUpdateOrderDraft use case

hichamboushaba reviewed (approved):
Nice work, it works well, I'm approving, but I have just two remarks for you before merging this:
1. This will make the payment section available even for M1, are we sure we'll be able to enable M2 by next release? otherwise is it OK to have an empty payment section visible for M1?
2. Something I noticed in the fees screen (forry for not noticing it in the fees screen review before), the spacing seems bigger than the other screens, because of applying a [padding](https://github.com/woocommerce/woocommerce-android/blob/5e2fcf1a461b384383ded2b4ad4e504c2e8049e6/WooCommerce/src/main/res/layout/fragment_order_creation_fee.xml#L16) + [margin](https://github.com/woocommerce/woocommerce-android/blob/5e2fcf1a461b384383ded2b4ad4e504c2e8049e6/WooCommerce/src/main/res/layout/fragment_order_creation_fee.xml#L31-L32), I think we should remove the `margins` for this case, WDYT? just for comparison, check for example how it looks vs shipping:

| Fees | Shipping |
| ----- | -------- |
| ![Screenshot_20220303_125906](https://user-images.githubusercontent.com/1657201/156561918-a46f3785-67e2-4c1f-8f89-29d4a70093af.png) | ![Screenshot_20220303_125927](https://user-images.githubusercontent.com/1657201/156561948-5a4b19bc-d612-4067-98da-7080de7d22b1.png) |

ThomazFB committed:
Merge branch 'trunk' into issue/5908-hide-fee-percentage

ThomazFB committed:
Make payment section visible from the start only in M2

ThomazFB commented:
Thanks once again for your review @hichamboushaba 🙇 

>This will make the payment section available even for M1, are we sure we'll be able to enable M2 by next release? otherwise is it OK to have an empty payment section visible for M1?

I don't think showing the payment section from the start on M1 would be necessary a problem, but I understand it doesn't need to be displayed either. I plan to submit the M2 flag removal today, but just to make this PR self-contained, I've adjusted the visibility rule, here's the fix: https://github.com/woocommerce/woocommerce-android/pull/5956/commits/9f2af9a63fa0e8a9382d594597b3294cb4beca4b.

>I think we should remove the margins for this case, WDYT?

Nice catch! I've adjusted the view to reduce the spacing between elements, but the switch component contains internal padding that creates some blank space around the view box. I tried some ways to work around it, but nothing that sounds a definitive solution. I'll try to adjust that in a separate PR so we can allow the improvements introduced here to move forward to tomorrow's release, WDYT? (unless you're aware of some simple and quick way to completely remove the internal spacing inside the SwitchMaterial)
<img width="228" alt="Screen Shot 2022-03-03 at 15 57 52" src="https://user-images.githubusercontent.com/5920403/156639987-bd71d87a-c605-4770-a943-5e7d143a4531.png">

hichamboushaba commented:
Thanks for the changes @ThomazFB 

> Nice catch! I've adjusted the view to reduce the spacing between elements, but the switch component contains internal padding that creates some blank space around the view box. I tried some ways to work around it, but nothing that sounds a definitive solution. I'll try to adjust that in a separate PR so we can allow the improvements introduced here to move forward to tomorrow's release, WDYT? (unless you're aware of some simple and quick way to completely remove the internal spacing inside the SwitchMaterial)

I think I wasn't clear on my remark, by `spacing` I meant the spacing around items, not between them, so the horizontal margin should be removed from the items, as it's already applied by the parent's padding, to clarify what I mean, check this patch:
```
Index: WooCommerce/src/main/res/layout/fragment_order_creation_fee.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/res/layout/fragment_order_creation_fee.xml b/WooCommerce/src/main/res/layout/fragment_order_creation_fee.xml
--- a/WooCommerce/src/main/res/layout/fragment_order_creation_fee.xml	(revision 989f1c388ea9702d680bce058786de5ff1fb3bf0)
+++ b/WooCommerce/src/main/res/layout/fragment_order_creation_fee.xml	(date 1646337567844)
@@ -19,7 +19,6 @@
                 android:id="@+id/fee_type_switch"
                 android:layout_width="wrap_content"
                 android:layout_height="wrap_content"
-                android:layout_marginEnd="@dimen/major_100"
                 android:text="@string/order_creation_fee_percentage_toggle_text"
                 app:layout_constraintEnd_toEndOf="parent"
                 app:layout_constraintTop_toTopOf="parent" />
@@ -28,7 +27,7 @@
                 android:id="@+id/fee_amount_edit_text"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:layout_marginHorizontal="@dimen/major_100"
+                android:layout_marginTop="@dimen/major_100"
                 android:hint="@string/order_creation_fee_amount_hint"
                 app:layout_constraintEnd_toEndOf="parent"
                 app:layout_constraintStart_toStartOf="parent"
@@ -40,7 +39,7 @@
                 android:id="@+id/fee_percentage_edit_text"
                 android:layout_width="match_parent"
                 android:layout_height="wrap_content"
-                android:layout_marginHorizontal="@dimen/major_100"
+                android:layout_marginTop="@dimen/major_100"
                 android:hint="@string/order_creation_fee_percentage_hint"
                 android:visibility="gone"
                 android:inputType="numberDecimal"
```

ThomazFB committed:
Adjust fee views horizontal margins

ThomazFB commented:
>by spacing I meant the spacing around items, not between them, so the horizontal margin should be removed from the items

Got it! Sorry for the confusion, it's fixed now here: https://github.com/woocommerce/woocommerce-android/pull/5956/commits/df787e58b5eada634c1a411058850f00b51955bc

<img src="https://user-images.githubusercontent.com/5920403/156652276-323fa50b-ace4-4c80-9537-f2d91aad26b0.png" width="540" height="1140" />

ThomazFB committed:
Merge branch 'trunk' into issue/5908-hide-fee-percentage

hichamboushaba commented:
Thanks @ThomazFB
